src/mixins/object_origin.mixin.ts

```diff
@@ -1,345 +1,404 @@
-
-(function (global) {
-  var fabric = global.fabric,
-    degreesToRadians = fabric.util.degreesToRadians,
-    originXOffset = {
-      left: -0.5,
-      center: 0,
-      right: 0.5,
-    },
-    originYOffset = {
-      top: -0.5,
-      center: 0,
-      bottom: 0.5,
+import { transformPoint } from '../util/misc/matrix';
+import { degreesToRadians } from '../util/misc/radiansDegreesConversion';
+import { CommonMethods } from './shared_methods.mixin';
+import { TDegree, TOriginX, TOriginY } from '../typedefs';
+import { Group } from '../shapes/group.class';
+import { sizeAfterTransform } from '../util/misc/objectTransforms';
+
+const originOffset = {
+  left: -0.5,
+  top: -0.5,
+  center: 0,
+  bottom: 0.5,
+  right: 0.5,
+};
+
+/**
+ * Resolves origin value relative to center
+ * @private
+ * @param {TOriginX | TOriginY} originValue originX / originY
+ * @returns number
+ */
+export const resolveOrigin = (
+  originValue: TOriginX | TOriginY | number
+): number =>
+  typeof originValue === 'string'
+    ? originOffset[originValue]
+    : originValue - 0.5;
+
+export class ObjectOrigin extends CommonMethods {
+  /**
+   * Top position of an object. Note that by default it's relative to object top. You can change this by setting originY={top/center/bottom}
+   * @type Number
+   * @default 0
+   */
+  top: number;
+
+  /**
+   * Left position of an object. Note that by default it's relative to object left. You can change this by setting originX={left/center/right}
+   * @type Number
+   * @default 0
+   */
+  left: number;
+
+  /**
+   * Object width
+   * @type Number
+   * @default
+   */
+  width: number;
+
+  /**
+   * Object height
+   * @type Number
+   * @default
+   */
+  height: number;
+
+  /**
+   * Object scale factor (horizontal)
+   * @type Number
+   * @default 1
+   */
+  scaleX: number;
+
+  /**
+   * Object scale factor (vertical)
+   * @type Number
+   * @default 1
+   */
+  scaleY: number;
+
+  /**
+   * Angle of skew on x axes of an object (in degrees)
+   * @type Number
+   * @default 0
+   */
+  skewX: number;
+
+  /**
+   * Angle of skew on y axes of an object (in degrees)
+   * @type Number
+   * @default 0
+   */
+  skewY: number;
+
+  /**
+   * Horizontal origin of transformation of an object (one of "left", "right", "center")
+   * See http://jsfiddle.net/1ow02gea/244/ on how originX/originY affect objects in groups
+   * @type String
+   * @default 'left'
+   */
+  originX: TOriginX;
+
+  /**
+   * Vertical origin of transformation of an object (one of "top", "bottom", "center")
+   * See http://jsfiddle.net/1ow02gea/244/ on how originX/originY affect objects in groups
+   * @type String
+   * @default 'top'
+   */
+  originY: TOriginY;
+
+  /**
+   * Angle of rotation of an object (in degrees)
+   * @type Number
+   * @default 0
+   */
+  angle: TDegree;
+
+  /**
+   * Width of a stroke used to render this object
+   * @type Number
+   * @default 1
+   */
+  strokeWidth: number;
```

Comment:
I know this class requires strokeWidth and strokeUniform but I think they belong in an abstract super class
Thoughts?
But we can move ahead and think about this in a follow up

Reply:
maybe yes i don't want to over engineer, no one is going to look at those classes and they are going to look at Object only or Object + interaction.

I want to put a target on eoy for finishing this thing and all the changes that do not touch the public api can be postponed for me.